_posts/2021-09-21-tweet-hashtags-relations.md

```diff
@@ -0,0 +1,59 @@
+
+	I actually found inserting data this way is not the most sustainable way to do it, as reconnecting to the DB is needed when any errors occurred before committing the changes to the database caused by using start_transaction(). 
+
+	An example code for building table 'tweets' is below:...
+
+	A screenshot of the tables:...
+
+4. **Calculate the Probability of One Hashtag Leading to Another Hashtag:**
+	Background Information: Two hashtags can be directly connected by appearing in the same tweet, or can be indirectly connected by connecting through one or more hashtags. In other words, in this example tweet "I love #dogs and #cats", #dogs and #cats are directly connected, because they appeared in the same tweet. If there is another tweet "#cats are the best #pets", this makes '#dogs' and '#pets' indirectly connected through '#cats'. For the simplicity of the project, I calculated the probability based only on the direct connection of two hashtags. 
+
+	The probability of #A leads to #B is calculated by MATH !!! dividing the count of #B that occurred with #A in the same tweet by the total count of hashtags in all of the tweets that #A occurred. Note: This formula makes the probability bidirectional (P(#A --> #B) != P(#B --> #A))
+
+	This calculation is done by joining a couple of SQL tables and the result is read into pandas dataframe, shown below:...
+
+5. **Visualizing the Probability (Hashtag Clustering Network):**
+	To make it easier to visualize the hashtag clustering network, only the hashtags having a sum(count) > 10 from all tweets are selected, making 166 hashtags are selected for visualization.
+
+	
+	Networkx (insert package!!!!) is used for visualization. 
+
+	To visualize how close two hashtags should be in the graph, the edge weight between two hashtags is used and calculated as the average of P(#A --> #B) and P(#B --> #A). ForceAtLas2 (insert package!!!!) is used for determining the node(hashtag) positions given the edge weight between two hashtags. In general, ForceAtLas2 makes two nodes with bigger weights closer to each other. 
+
+	community_louvain (insert package!!!!) is used for clustering based on how close the nodes are to each other. 
+
+	There are 20 clusters in the graph. To help with visualization, node size is determined by the number of hashtags of all tweets (bigger size meaning more number). Weight edge line transparency is determined by the weight (more transparent meaning smaller weight/closer).
+
+	The graph is suggesting hashtags having similar meaning or in similar categories are more likely to lead to each other (being grouped together in the graph). For example, #twitch is clustered with #gaming, #gamer, #nintendoswitch and etc. #covid19, #covid and #getvaccinated are grouped together. Bitcoin hashtags are also grouped together. This result perhaps indicates including popular hashtags with more relevance to the new hashtags can help increase the popularity of the new hashtags.
```

Comment:
You should also mention the interesting fact we found, that hashtags which are connected in the graph, can sometimes have polar opposite connotation (such as the pro vs. anti biden hashtags)